File: app/src/main/java/com/elishaazaria/sayboard/ime/recognizers/providers/VoskLocalProvider.java
package com.elishaazaria.sayboard.ime.recognizers.providers;

import com.elishaazaria.sayboard.Tools;
import com.elishaazaria.sayboard.data.LocalModel;
import com.elishaazaria.sayboard.data.LocalModelType;
import com.elishaazaria.sayboard.ime.IME;
import com.elishaazaria.sayboard.ime.recognizers.RecognizerSource;
import com.elishaazaria.sayboard.ime.recognizers.VoskLocal;
import com.elishaazaria.sayboard.ime.recognizers.WhisperLocal;

import java.util.List;

public class VoskLocalProvider implements RecognizerSourceProvider {
    private final IME ime;

    public VoskLocalProvider(IME ime) {
        this.ime = ime;
    }

    @Override
    public void loadSources(List<RecognizerSource> recognizerSources) {
        for (LocalModel localModel : Tools.getInstalledModelsList(ime)) {
            if (localModel.modelType == LocalModelType.VOSK) {
                recognizerSources.add(new VoskLocal(localModel));
            } else {
                recognizerSources.add(new WhisperLocal(localModel));
            }
        }
    }
}

File: app/src/main/java/com/volla/vollaboard/ime/recognizers/providers/VoskLocalProvider.kt
package com.volla.vollaboard.ime.recognizers.providers

import android.content.Context
import com.volla.vollaboard.Constants
import com.volla.vollaboard.Tools
import com.volla.vollaboard.data.InstalledModelReference
import com.volla.vollaboard.data.ModelType
import com.volla.vollaboard.ime.recognizers.RecognizerSource
import com.volla.vollaboard.ime.recognizers.VoskLocal
import java.util.Locale

class VoskLocalProvider(private val context: Context) : RecognizerSourceProvider {
    override fun getInstalledModels(): List<InstalledModelReference> {
        val models: MutableList<InstalledModelReference> = ArrayList()
        val modelsDir = Constants.getModelsDirectory(context)
        if (!modelsDir.exists()) return models
        for (localeFolder in modelsDir.listFiles()!!) {
            if (!localeFolder.isDirectory) continue
            val locale = Locale.forLanguageTag(localeFolder.name)
            for (modelFolder in localeFolder.listFiles()!!) {
                if (!modelFolder.isDirectory) continue
//                val name = modelFolder.name
                val model = InstalledModelReference(
                    modelFolder.absolutePath,
                    locale.displayName,
                    ModelType.VoskLocal
                )
                models.add(model)
            }
        }
        return models
    }

    override fun recognizerSourceForModel(localModel: InstalledModelReference): RecognizerSource? {
        return VoskLocal(Tools.getVoskModelFromReference(localModel) ?: return null)
    }
}